pkg/predicate/predicate.go

```diff
@@ -98,22 +98,29 @@ func (p Funcs) Generic(e event.GenericEvent) bool {
 	return true
 }
 
+// FilterFunc implements Predicate.
+type FilterFunc[T client.Object] func(T) bool
+
 // NewPredicateFuncs returns a predicate funcs that applies the given filter function
 // on CREATE, UPDATE, DELETE and GENERIC events. For UPDATE events, the filter is applied
 // to the new object.
-func NewPredicateFuncs(filter func(object client.Object) bool) Funcs {
+func NewPredicateFuncs[T client.Object](filter FilterFunc[T]) Funcs {
 	return Funcs{
 		CreateFunc: func(e event.CreateEvent) bool {
-			return filter(e.Object)
+			v := e.Object.(T)
```

Comment:
We still have the type assertion here and its possible to misuse this by constructing a predicate for the wrong type. What I'd want is the whole chain of source/predicate/handler using generics. I'll update the issue to clarify that

Reply:
On it :smile:

Reply:
#### Update

The only thing I'm having trouble with is builder
Right now, builder can watch multiple resources and I can't really instantiate many watches with different type parameters using generics without providing a top level function and potentially changing the interface of the `Watch` method
I'm looking into ways of doing it, so the change is as minimal as possible

Reply:
I looked into this some time ago. Golang doesn't support generic methods, which is why we can not keep assembling everything in the controller.Watch but instead have to provide predicate+handler in a constructor to the source